### PR TITLE
Chapter 3 - ILP - Dynamic Scheduling

### PR DESCRIPTION
**Pull request by pveentjer**
This is work in progress.

### CODE REVIEW

File: chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md [hunk 1]
```diff
@@ -1,27 +1,91 @@
...
+
+It gets worse if the loads miss in the cache because the CPU could stall even though the process remains scheduled on the core; so you get a misleading high CPU utilization. That is why it is good to check the instructions per cycle (IPC) to see if the core is doing anything useful. When there are multiple independent dependency chains, then there is still a chance that the CPU can find ILP. 
+
+The OoOE in the Tomasulo algorithm is implemented using reservation stations where instructions wait for their operands to become available. Once they are available, the instructions can be dispatched to the approproate execution unit. 
+
+Most implementations strive to balance the hardware cost with the potential return. Typically, the size of the reorder buffer determines how far ahead the hardware can look for scheduling such independent instructions. With the introduction of the Apple M1, the reorder buffer size almost doubled compared to the fastest contemporary X86 implementations and Apple showed there is a lot of potential to uncover more parallel instructions inside a singe instruction stream. Modern processors will be wider (more execution units) and deeper (larger rob).
```
Comment (jerrinot) [inline chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md:75]: this paragraph mentions ROB - line 75.
but ROB is only introduced further down, at line 115. 

(too much out of order? :))

**Comment by dendibakh** [comment]
@pveentjer , I made a few changes.
1) I moved your register renaming examples up in the section "3.2 Pipelining"
2) I sticked to using OOO, because I refer to it later in many places in the book, for example, "... OOO engine ..."
3) I feel like this is not the right place to talk about memory ordering. I would rather save it for section 3.8.3 Load-Store Unit or chapter 12 (section Architecture-Specific Optimizations).
4) Also, please keep in mind, this is mostly theoretical discussion about ideas. I'm OK with keeping things abstract here and not get into details. I have a big section that exemplifies a real implementation: section "3.8 Modern CPU Design". That is where we can discuss nuances. BTW, maybe I can come up with a better name for this section...
5) I left some TODO in the text, please check them.

File: chapters/3-CPU-Microarchitecture/3-2 Pipelining.md [hunk 2]
```diff
@@ -38,13 +38,22 @@ In real implementations, pipelining introduces several constraints that limit th
   A *write-after-read* (WAR) hazard requires a dependent write to execute after a read. It occurs when instruction `x+1` writes a source before instruction `x` reads the source, resulting in the wrong new value being read. A WAR hazard is not a true dependency and is eliminated by a technique called [register renaming](https://en.wikipedia.org/wiki/Register_renaming).[^1] It is a technique that abstracts logical registers from physical registers. CPUs support register renaming by keeping a large number of physical registers. Logical (architectural) registers, the ones that are defined by the ISA, are just aliases over a wider register file. With such decoupling of [architectural state](https://en.wikipedia.org/wiki/Architectural_state),[^3] solving WAR hazards is simple: we just need to use a different physical register for the write operation. For example:
```
Comment (pveentjer) [inline chapters/3-CPU-Microarchitecture/3-2 Pipelining.md:39]: I would move all the register renaming logic to the 3.3 chapter. Register renaming AFAIK was used first in the Tomasulo algorithm and one can use pipelining without register renaming.

File: chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md [hunk 3]
```diff
@@ -1,27 +1,55 @@
...
+[TODO]:
+Peter: "(not true: on ARM the OOoE of loads/stores is allowed to become visible.. loads for sure. So it depends on the memory model of the ISA). "
+Denis: I'm not familiar. What are the conditions in which this could happen? Is it about relaxed consistency?
```
Comment (pveentjer) [inline chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md:11]: Yes. So it could happen that an earlier load runs into a cache miss and a later load doesn't and then the 2 loads don't observe the stores in program order. On X86 this is prohibited since loads can't be reordered with other loads, but on the ARM this would be perfectly fine. On ARM a LDR doesn't have any memory ordering semantics and on x86 a load has acquire semantics.

File: chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md [hunk 4]
```diff
@@ -1,27 +1,55 @@
...
+
+[TODO]: Peter, I guess you could expand here: "It depends on the ISA memory order which reordering is allowed to become architecturally visible."
 
-### Superscalar Engines and VLIW
+The OOO execution in the Tomasulo algorithm is implemented using the Reorder Buffer (ROB) and Reservation Station (RS). The ROB is a circular buffer that keeps track of the state of each instruction, and in modern processors it has more than a thousand entries. Typically, the size of the ROB determines how far ahead the hardware can look for scheduling such independent instructions. Instructions are inserted in the ROB in program order, can execute out of order, and retire in program order. Register renaming is done when the instructions are placed in the ROB.
```
Comment (pveentjer) [inline chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md:37]: A single reservation station is needed for a single instruction. So a CPU needs to have a large number of reservation stations.

Comment (pveentjer) [reply chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md:37]: I don't know of any processor that has 1000 entries in the ROB. The Apple M1 has something like 630 and that was double that of Intel/AMD. I think it is safer to say that the rob capacities is in the hundreds on modern processors.

File: chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md [hunk 5]
```diff
@@ -1,27 +1,55 @@
...
+
+[TODO]: Peter, I guess you could expand here: "It depends on the ISA memory order which reordering is allowed to become architecturally visible."
 
-### Superscalar Engines and VLIW
+The OOO execution in the Tomasulo algorithm is implemented using the Reorder Buffer (ROB) and Reservation Station (RS). The ROB is a circular buffer that keeps track of the state of each instruction, and in modern processors it has more than a thousand entries. Typically, the size of the ROB determines how far ahead the hardware can look for scheduling such independent instructions. Instructions are inserted in the ROB in program order, can execute out of order, and retire in program order. Register renaming is done when the instructions are placed in the ROB. 
+
+[TODO]: Peter, is the term precise exceptions a conventional one?
```
Comment (pveentjer) [inline chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md:39]: AFAIK it is a term that is well established.

File: chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md [hunk 6]
```diff
@@ -1,27 +1,55 @@
...
+
+From the ROB, instructions are inserted in the RS, which has much fewer entries. Once instructions are in the RS, they wait for their input operands to become available. When inputs are available, instructions can be dispatched to the appropriate execution unit. So instructions can be executed in any order once their operands become available and are not tied to the program order any longer. Modern processors are becoming wider (can execute many instructions in one cycle) and deeper (larger ROB, RS, and other buffers), which demonstrates that there is a lot of potential to uncover more ILP in production applications.
```
Comment (pveentjer) [inline chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md:43]: The reservation station can only deal with a single instruction. So there will be many reservation stations.

File: chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md [hunk 7]
```diff
@@ -48,7 +76,19 @@ The concept of speculative execution.
...
+Even though modern processors retire the ROB in order, it doesn't mean that OoOE execution can't be observed. Loads could be performed out of order, and even though stores typically end up in a store buffer, the stores don't need to be committed to the coherent cache in order. It depends on the memory model of the ISA which reorderings are allowed to become architecturally visible.
+
+[TODO]: please define TSO
+
+It isn't only branches the CPU can speculate on. Another example is speculating on reordering of loads in case of the x86. According to the TSO memory model, the memory model of the x86, loads need to be performed in program order. But due to OOO execution, they can execute out of order which will help to prevent stalls due to cache misses. When a core detects that a store issued by another core could potentially lead to a violation of the TSO, a speculation failure is triggered [^7].
+
+[TODO]: [... a speculation failure is triggered] and then what happens next? The load gets reissued?
```
Comment (pveentjer) [inline chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md:91]: Correct. Effectively the same behavior as a branch misprediction. So the pipeline restarts, the loads get executed again and hopefully this time no conflicting write will be detected.

File: chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md [hunk 8]
```diff
@@ -1,27 +1,55 @@
...
+The pipeline example in Figure @fig:Pipelining shows all instructions moving through the different stages of the pipeline in-order, i.e., in the same order as they appear in the program, also know as *program order*. Most modern CPUs support *out-of-order* (OOO) execution, where sequential instructions can enter the execution pipeline stage in any arbitrary order only limited by their dependencies and resources availability. CPUs with OOO execution must still give the same result as if all instructions were executed in the program order
```
Comment (pveentjer) [inline chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md:7]: I would rename the 'pipeline' from 'execution pipeline stage'.

File: chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md (hunk 4 above)
Comment (pveentjer) [reply chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md:37]: I think the explanation is now out of order :) The ROB isn't part of the original Tomasulo algorithm. It is an extension to support speculative execution and it isn't needed for out of order execution.

File: chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md [hunk 9]
```diff
@@ -1,27 +1,55 @@
...
+
+To eliminate false dependencies, the Tomasulo algorithm makes use of register renaming, which we discussed in the previous section. Because of that, performance is greatly improved compared to scoreboarding. However, a sequence of instructions with true dependencies (RAW), also known as *dependency chain*, are still problematic for OOO execution, because there is no increase in the ILP after register renaming since all the RAW dependencies are preserved. Dependency chains are often found in loops (loop carried dependency) where current loop iteration depends on the results produced on the previous iteration. It gets even worse if a number of instructions depend on a load that missed in caches. In that case, execution would stall waiting for that data to be fetched from memory.
```
Comment (pveentjer) [inline chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md:33]: A sequence of instructions with true dependencies doesn't need to be a problem as long as the dependency isn't 'carried'. So the steps in the sequence are not dependent on each other.

I would rephrase it like this:
```
However, a sequence of instructions that carry a RAW dependency, also known as *dependency chain* ....
```

This way it is more clear that there is a dependency between the steps.

File: chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md [hunk 10]
```diff
@@ -48,7 +76,19 @@ The concept of speculative execution.
...
+An instruction that is executed speculatively is marked as such in the ROB. Once it is not speculative any longer, it can retire in program order. Here is where the architectural state is committed, and architectural registers are updated. Because results of the speculative instructions are not committed, it is easy to roll back when a misprediction happens.
+
+[TODO]: it's hard to understand the following sentence without an example.
+It depends on the microarchitecture if the rollback happens on all speculative state of the ROB, or only till the instruction where the speculation failed. The latter makes the misprediction a lot less expensive.
+
+[TODO]: it's hard to understand the following sentence without an example. We haven't introduced the store buffer yet. I suggest to keep it for either section 3.8.3 Load-Store Unit or chapter 12 (section Architecture-Specific Optimizations).
```
Comment (pveentjer) [inline chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md:84]: I agree. Just delete the text and I'll do as you suggested.

File: chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md [hunk 11]
```diff
@@ -48,7 +76,19 @@ The concept of speculative execution.
...
+Even though modern processors retire the ROB in order, it doesn't mean that OoOE execution can't be observed. Loads could be performed out of order, and even though stores typically end up in a store buffer, the stores don't need to be committed to the coherent cache in order. It depends on the memory model of the ISA which reorderings are allowed to become architecturally visible.
+
+[TODO]: please define TSO
```
Comment (pveentjer) [inline chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md:87]: Perhaps better to move this whole section to  3.8.3/12. Better to delete it for now so that only the 3.3 chapter is changed in this PR.

File: chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md (hunk 3 above)
Comment (pveentjer) [reply chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md:11]: We should move the whole discussion about memory ordering to its own place. 

If you change this

`
CPUs with OOO execution must still give the same result as if all instructions were executed in the program order`

To this

`
CPUs with OOO execution should not be able to observe the out of order execution of its own instructions.
`

Then it is more clear that the OOO should not influence the CPU in isolation.

File: chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md (hunk 7 above)
Comment (pveentjer) [reply chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md:91]: Perhaps better to simplify it to:
`
It isn't only branches the CPU can speculate on, it could for example also speculate that the memory order isn't violated. For more info see ...
`

File: chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md [hunk 12]
```diff
@@ -75,6 +115,7 @@ Another common technique used is called hybrid prediction. The idea is that some
...
 
-[^4]: Tomasulo algorithm - [https://en.wikipedia.org/wiki/Tomasulo_algorithm](https://en.wikipedia.org/wiki/Tomasulo_algorithm).
-[^5]: Scoreboarding - [https://en.wikipedia.org/wiki/Scoreboarding](https://en.wikipedia.org/wiki/Scoreboarding).
-[^6]: 5th Championship Branch Prediction competition - [https://jilp.org/cbp2016](https://jilp.org/cbp2016).
\ No newline at end of file
+[^4]: Scoreboarding - [https://en.wikipedia.org/wiki/Scoreboarding](https://en.wikipedia.org/wiki/Scoreboarding).
+[^5]: Tomasulo algorithm - [https://en.wikipedia.org/wiki/Tomasulo_algorithm](https://en.wikipedia.org/wiki/Tomasulo_algorithm).
+[^6]: 5th Championship Branch Prediction competition - [https://jilp.org/cbp2016](https://jilp.org/cbp2016).
+[^7]: Machine Clears - [https://portal.nacad.ufrj.br/online/intel/vtune2017/help/GUID-F0FD7660-58B5-4B5D-AA9A-E1AF21DDCA0E.html](https://portal.nacad.ufrj.br/online/intel/vtune2017/help/GUID-F0FD7660-58B5-4B5D-AA9A-E1AF21DDCA0E.html).
```
Comment (pveentjer) [inline chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md:121]: Machine clear can be moved to right section together with the memory ordering part.

File: chapters/3-CPU-Microarchitecture/3-2 Pipelining.md (hunk 2 above)
Comment (dendibakh) [reply chapters/3-CPU-Microarchitecture/3-2 Pipelining.md:39]: Let me think more about it, but I tend to think that it's good already.

File: chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md (hunk 4 above)
Comment (dendibakh) [reply chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md:37]: > I don't know any processor that has 1000 entries in the ROB. The Apple M1 has something like 630 and that was double that of Intel/AMD. It is safer to say that the ROB capacity is in the hundreds on modern processors.

I'm going ahead of myself and I need to sleep more. :) Just checked chips-and-cheese and yes, you're right. Will change.

> A single reservation station is needed for a single instruction. So a CPU needs to have a large number of reservation stations.

I think this is not true. Again, from chips and cheese:

image

> I think the explanation is now out of order :) The ROB isn't part of the original Tomasulo algorithm. It is an extension to support speculative execution and it isn't needed for out of order execution.

I'm not very well versed in the history here, but I think you still need ROB to retire instructions in order, no? We can say something like "Modern processors implement derivatives of Tomasulo's algorithm that include ROB and RS ..."

File: chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md (hunk 6 above)
Comment (dendibakh) [reply chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md:43]: See previous discussion.

File: chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md (hunk 9 above)
Comment (dendibakh) [reply chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md:33]: Agree.

File: chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md (hunk 7 above)
Comment (dendibakh) [reply chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md:91]: I removed it for now. I think there is no need for nuances here. Let's explain it elsewhere. Otherwise, readers will be overwhelmed with complex information and will context-switch from branch mispredictions to memory ordering to branch mispredictions again.

File: chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md (hunk 3 above)
Comment (dendibakh) [reply chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md:11]: I would rather leave it as it is. Let's not overwhelm readers with nuances, they would second-guess themselves whether they really understand it or not. We will have a chance to correct ourselves later.

**Comment by dendibakh** [comment]
I cleaned up the PR, @pveentjer please let me know if you have further comments.

File: chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md (hunk 4 above)
Comment (pveentjer) [reply chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md]: To be honest; I'm also slightly confused about how the original Tomasulo deals with in-order retirement.

**Comment by pveentjer** [comment]
It looks good to me. Thank you for the review!

File: chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md [hunk 13]
```diff
@@ -1,27 +1,47 @@
...
+
+#### Dynamic scheduling
+
+To overcome the problem with static scheduling, modern processors use dynamic scheduling. The two most important algorithms for dynamic scheduling are [Scoreboading](https://en.wikipedia.org/wiki/Scoreboarding),[^4] and the [Tomasulo algorithm](https://en.wikipedia.org/wiki/Tomasulo_algorithm).[^5]
+
+Scoreboarding was first implemented in the CDC6600 in 1960s. Its main drawback is that it not only preserves true dependencies (RAW), but also false dependencies (WAW and WAR), and therefore it provides suboptimal ILP. False dependencies are caused by the small number of architectural registers, which is typically between 16 and 32 in modern ISAs. That is why all modern processors have adopted the Tomasulo algoritm for dynamic scheduling. Tomasolu algorithm was invented in 1960s by Robert Tomasulo and first implemented in the IBM360 model 91.
```
Comment (cf-natali) [inline chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md:27]: Typo: "Tomasulu algorithm".

Comment (cf-natali) [inline chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md:27]: Typo: "algoritm".

File: chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md [hunk 14]
```diff
@@ -1,27 +1,47 @@
...
+
+From the ROB, instructions are inserted in the RS, which has much fewer entries. Once instructions are in the RS, they wait for their input operands to become available. When inputs are available, instructions can be dispatched to the appropriate execution unit. So instructions can be executed in any order once their operands become available and are not tied to the program order any longer. Modern processors are becoming wider (can execute many instructions in one cycle) and deeper (larger ROB, RS, and other buffers), which demonstrates that there is a lot of potential to uncover more ILP in production applications.
```
Comment (cf-natali) [inline chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md:35]: "When inputs are available, instructions can be dispatched to the appropriate execution unit."

It would be better to use "issued" here.
Intel uses "issue" and "dispatch" in the opposite way of everyone else, so I think sticking with the generally accepted meaning (in this case "issue" for submission to an execution unit) would be better.

Comment (dendibakh) [reply chapters/3-CPU-Microarchitecture/3-3 Exploiting ILP.md:35]: Thank you @cf-natali for the review!

**Comment by dendibakh** [comment]
@pveentjer , thanks a lot!
